File: plugins/plugin-agent-orchestrator/rust/src/providers/task_context.rs
//! Task context provider for the Agent Orchestrator plugin.

use crate::service::{AgentOrchestratorService, Runtime};
use serde_json::json;

/// Provider definition
pub const TASK_CONTEXT_PROVIDER: ProviderDef = ProviderDef {
    name: "TASK_CONTEXT",
    description: "Provides context about active and recent orchestrated tasks",
    position: 90,
};

/// Provider definition
#[derive(Debug, Clone, Copy)]
pub struct ProviderDef {
    pub name: &'static str,
    pub description: &'static str,
    pub position: i32,
}

/// Get task context from the service
pub async fn get_task_context<R: Runtime>(
    service: &AgentOrchestratorService<R>,
) -> serde_json::Value {
    let context_text = match service.get_task_context().await {
        Ok(text) => text,
        Err(_) => "Task orchestrator service error".to_string(),
    };

    let current = service.get_current_task().await.ok().flatten();
    let task_count = service.get_tasks().await.map(|t| t.len()).unwrap_or(0);

    json!({
        "values": {
            "taskContext": context_text,
            "currentTaskName": current.as_ref().map(|t| &t.name).unwrap_or(&"None".to_string()),
            "currentTaskStatus": current.as_ref().map(|t| t.metadata.status.to_string()).unwrap_or_else(|| "N/A".to_string()),
        },
        "text": format!("# Task Context\n\n{}", context_text),
        "data": {
            "taskCount": task_count,
            "currentTaskId": current.as_ref().map(|t| &t.id),
        }
    })
}

/// Add header to content
pub fn add_header(title: &str, content: &str) -> String {
    format!("{}\n\n{}", title, content)
}

File: plugins/plugin-agent-orchestrator/typescript/src/providers/task-context.ts
import type { IAgentRuntime, Memory, Provider, ProviderResult, State } from "@elizaos/core";
import type { AgentOrchestratorService } from "../services/agent-orchestrator-service.js";

async function getService(runtime: IAgentRuntime): Promise<AgentOrchestratorService | null> {
  return (await runtime.getService("CODE_TASK")) as AgentOrchestratorService | null;
}

function formatStatus(status: string): string {
  switch (status) {
    case "pending":
      return "⏳ pending";
    case "running":
      return "🔄 running";
    case "paused":
      return "⏸️ paused";
    case "completed":
      return "✅ completed";
    case "failed":
      return "❌ failed";
    case "cancelled":
      return "🛑 cancelled";
    default:
      return status;
  }
}

export const taskContextProvider: Provider = {
  name: "TASK_CONTEXT",
  description: "Summary of orchestrated tasks and current selection",
  dynamic: true,
  get: async (
    runtime: IAgentRuntime,
    _message: Memory,
    _state?: State,
  ): Promise<ProviderResult> => {
    const svc = await getService(runtime);
    if (!svc) {
      return { text: "Task service not available." };
    }

    const current = await svc.getCurrentTask();
    const tasks = await svc.getRecentTasks(10);

    if (tasks.length === 0) {
      return { text: "No tasks have been created yet." };
    }

    const lines: string[] = [];
    if (current) {
      lines.push(`## Current Task`);
      lines.push(`- Name: ${current.name}`);
      lines.push(`- Status: ${formatStatus(current.metadata.status)}`);
      lines.push(`- Progress: ${current.metadata.progress}%`);
      lines.push(`- Provider: ${current.metadata.providerLabel ?? current.metadata.providerId}`);
      lines.push("");
    }

    lines.push("## Recent Tasks");
    for (const t of tasks) {
      const marker = current?.id === t.id ? " (current)" : "";
      lines.push(
        `- ${t.name} — ${formatStatus(t.metadata.status)} ${t.metadata.progress}%${marker}`,
      );
    }

    return { text: lines.join("\n").trim() };
  },
};
